### PR TITLE
First pass at whitelist for DNS records

### PR DESCRIPTION
**Pull request by mstump**
We've got multiple clusters sharing the same route53 zone. It was way too easy to accidentally create resources on one cluster that stole records from another. We wanted to be able to whitelist DNS records prevent accidental stealing or allocation of records.

### CODE REVIEW

File: dns-controller/pkg/dns/zonespec.go [hunk 1]
```diff
@@ -40,26 +47,36 @@ func ParseZoneSpec(s string) (*ZoneSpec, error) {
...
 	// example.com/1234: Match by name & id
-	return &ZoneSpec{Name: name, ID: tokens[1]}, nil
+	return &ZoneSpec{Name: name, Pattern: glob.MustCompile(name, '.'), ID: tokens[1]}, nil
+}
+
+func ParseRecordWhiteList(s string) (*RecordWhiteList, error) {
+	s = strings.TrimSpace(s)
+	g, err := glob.Compile(s, '.')
+	if err != nil {
+		return nil, err
```
Comment (chrislovecnm) [inline dns-controller/pkg/dns/zonespec.go]: Need a fmt.Errorf here since this is an external library.

File: dns-controller/pkg/dns/zonespec.go [hunk 2]
```diff
@@ -73,19 +90,32 @@ func ParseZoneRules(zones []string) (*ZoneRules, error) {
...
+		r.Catchall = true
+	}
+
+	for _, s := range whitelists {
+		whiteList, err := ParseRecordWhiteList(s)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing %q: %v", s, err)
+		}
+
+		r.WhiteLists = append(r.WhiteLists, whiteList)
```
Comment (chrislovecnm) [inline dns-controller/pkg/dns/zonespec.go]: Probably good to log that we have a whitelist

File: dns-controller/pkg/dns/zonespec.go [hunk 3]
```diff
@@ -73,19 +90,32 @@ func ParseZoneRules(zones []string) (*ZoneRules, error) {
...
+	}
+
+	if len(whitelists) == 0 {
+		glog.Infof("No whitelist rules specified, will permit management of any record")
 	}
```
Comment (chrislovecnm) [inline dns-controller/pkg/dns/zonespec.go]: glog.V(2) please

**Comment by mstump** [comment]
upstream doesn't have glob, my reasoning for using it instead of regex is that globs are easier to reason about and whitelists do have a security implication. Downside being that we're pulling in another dependency which is a PITA. REGEX vs another dep: thoughts?

Haven't addressed the bazel issue yet.

**Comment by chrislovecnm** [comment]
@mstump how complex is the regular expressions.  Need more context for a good decision.  I tend to like using libraries, but kinda on the fence.

You still have protokube issues and the integration with bootstrapchannel builder that I mentioned.

**Pull request closed without merging**